File: libraries/ndf/ndf_gtune.f
      SUBROUTINE NDF_GTUNE( TPAR, VALUE, STATUS )
*+
*  Name:
*     NDF_GTUNE

*  Purpose:
*     Obtain the value of an NDF_ system tuning parameter.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF_GTUNE( TPAR, VALUE, STATUS )

*  Description:
*     The routine returns the current value of an NDF_ system internal
*     tuning parameter.

*  Arguments:
*     TPAR = CHARACTER * ( * ) (Given)
*        Name of the tuning parameter whose value is required (case
*        insensitive). This name may be abbreviated, to no less than 3
*        characters.
*     VALUE = INTEGER (Returned)
*        Value of the parameter.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     See the NDF_TUNE routine for a list of the tuning parameters
*     currently available.

*  Copyright:
*     Copyright (C) 1993 Science & Engineering Research Council
*     Copyright (C) 2007-2009 Science & Technology Facilities Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     DSB: David S. Berry (JACH, UClan)
*     {enter_new_authors_here}

*  History:
*     4-OCT-1991 (RFWS):
*        Original version.
*     14-OCT-1991 (RFWS):
*        Report contextual error information.
*     17-OCT-1991 (RFWS):
*        Fixed bug: missing argument to ERR_REP.
*     5-NOV-1993 (RFWS):
*        Added new tuning parameters to control foreign format
*        conversion.
*     11-MAR-1997 (RFWS):
*        Add the DOCVT tuning parameter.
*     1-NOV-2007 (DSB):
*        Add the PXT... tuning parameters.
*     18-SEP-2009 (DSB):
*        Add the AUTO_HISTORY tuning parameter.
*     16-JUL-2012 (DSB):
*        Add the SECMAX tuning parameter.
*     10-AUG-2018 (DSB):
*        Add the FIXDT tuning parameter.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'AST_PAR'          ! AST_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_CONST'        ! NDF_ private constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Global Variables:
      INCLUDE 'NDF_TCB'          ! NDF_ Tuning Control Block
*        TCB_AUTOHISTORY = LOGICAL (Read)
*           Automatic History creation flag.
*        TCB_DOCVT = LOGICAL (Read)
*           Do format conversions flag.
*        TCB_ETFLG = LOGICAL (Read)
*           Error tracing flag.
*        TCB_KEEP = LOGICAL (Read)
*           Keep NDF data objects flag.
*        TCB_SECMAX = INTEGER (Read)
*           Maximum no. of mega-pixels in a section.
*        TCB_SHCVT = LOGICAL (Read)
*           Show format conversions flag.
*        TCB_WARN = LOGICAL (Read)
*           Warning message flag.
*        TCB_PXT = INTEGER (Read)
*           An AST pointer to a KeyMap holding the names of NDF
*           extensions and their associated default propagation flags.
*        TCB_FIXDT = LOGICAL (Read)
*           Use a fixed date and time in history records.

*  Arguments Given:
      CHARACTER * ( * ) TPAR

*  Arguments Returned:
      INTEGER VALUE

*  Status:
      INTEGER STATUS             ! Global status

*  External References:
      LOGICAL NDF1_SIMLR         ! String compare with abbreviation

*  Local Variables:
      CHARACTER XNAME*(DAT__SZNAM) ! An NDF extension name

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Ensure that the TCB is initialised.
      CALL NDF1_INTCB( STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  Test the tuning parameter name supplied against each permitted value
*  in turn, allowing abbreviation...

*  Error tracing flag.
*  ==================
*  If TRACE was specified, then return the error tracing flag value.
         IF ( NDF1_SIMLR( TPAR, 'TRACE', NDF__MINAB ) ) THEN
            IF ( TCB_ETFLG ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Do format conversion flag.
*  =========================
*  If DOCVT was specified, then return the do format conversion flag
*  value.
         ELSE IF ( NDF1_SIMLR( TPAR, 'DOCVT', NDF__MINAB ) ) THEN
            IF ( TCB_DOCVT ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Keep NDF objects flag.
*  =====================
*  If KEEP was specified, then return the keep NDF objects flag value.
         ELSE IF ( NDF1_SIMLR( TPAR, 'KEEP', NDF__MINAB ) ) THEN
            IF ( TCB_KEEP ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Create History component automatically flag.
*  ============================================
         ELSE IF ( NDF1_SIMLR( TPAR, 'AUTO_HISTORY', NDF__MINAB ) ) THEN
            IF ( TCB_AUTOHISTORY ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Show data conversion flag.
*  =========================
*  If SHCVT was specified, then return the show data conversion flag
*  value.
         ELSE IF ( NDF1_SIMLR( TPAR, 'SHCVT', NDF__MINAB ) ) THEN
            IF ( TCB_SHCVT ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Warning message flag.
*  ====================
*  If WARN was specified, then return the warning message flag value.
         ELSE IF ( NDF1_SIMLR( TPAR, 'WARN', NDF__MINAB ) ) THEN
            IF ( TCB_WARN ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Fix History date/time flag.
*  ===========================
*  If FIXDT was specified, then return the fixed history date/time flag value.
         ELSE IF ( NDF1_SIMLR( TPAR, 'FIXDT', NDF__MINAB ) ) THEN
            IF ( TCB_FIXDT ) THEN
               VALUE = 1
            ELSE
               VALUE = 0
            END IF

*  Extension propagation.
*  ======================
*  Any tuning flag that begins with "PXT" is assumed to be terminated
*  with the name of an extension. The tuning flag value is non-zero if
*  the extension should be propagated by default by NDF_PROP, etc, and
*  is zero if the extension should not be propagated by default.
         ELSE IF ( NDF1_SIMLR( TPAR(1:3), 'PXT', NDF__MINAB ) ) THEN

* Get the extension name and check it is not blank.
            XNAME = TPAR( 4 : )
            IF( XNAME .NE. ' ' ) THEN

*  If there is a KeyMap, and it contains a value for the named extension,
*  use the stored value. Otherwise use a defualt value of 1 (all
*  extensions are propagated by default).
               IF( TCB_PXT .NE. AST__NULL ) THEN
                  IF( .NOT. AST_MAPGET0I( TCB_PXT, XNAME, VALUE,
     :                                    STATUS ) ) VALUE = 1
               ELSE
                  VALUE = 1
               END IF

*  If the extension name was blank, then report an error.
            ELSE
               STATUS = NDF__TPNIN
               CALL ERR_REP( 'NDF_GTUNE_WARN',
     : '''PXT'' is not a valid tuning parameter name - it should '//
     : 'be followed by an NDF extension name (possible programming '//
     : 'error).', STATUS )
            END IF

*  Maximum size of a section.
*  ==========================
         ELSE IF ( NDF1_SIMLR( TPAR, 'SECMAX', NDF__MINAB ) ) THEN
            VALUE = TCB_SECMAX

*  Unknown tuning parameter.
*  ========================
*  Report an error if the tuning parameter name is not recognised.
         ELSE
            STATUS = NDF__TPNIN
            CALL MSG_SETC( 'TPAR', TPAR )
            CALL ERR_REP( 'NDF_GTUNE_TPAR',
     : '''^TPAR'' is not a valid tuning parameter name (possible ' //
     : 'programming error).',
     :                    STATUS )
         END IF
      END IF

*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'NDF_GTUNE_ERR',
     : 'NDF_GTUNE: Error obtaining the value of an NDF_ system ' //
     : 'tuning parameter.',
     :                 STATUS )
         CALL NDF1_TRACE( 'NDF_GTUNE', STATUS )
      END IF

      END

File: libraries/ndf/ndfGtune.c
#include "sae_par.h"
#include "dat_par.h"
#include "star/util.h"
#include "ndf_ast.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "ndf.h"
#include "star/util.h"
#include <string.h>
#include "mers.h"

void ndfGtune_( const char *tpar, int *value, int *status ){
/*
*+
*  Name:
*     ndfGtune

*  Purpose:
*     Obtain the value of an NDF_ system tuning parameter.

*  Invocation:
*     void ndfGtune( const char *tpar, int *value, int *status )

*  Description:
*     This function returns the current value of an NDF_ system internal
*     tuning parameter.

*  Parameters:
*     tpar
*        Pointer to a null terminated string holding the name of the tuning
*        parameter whose value is required (case insensitive). This name
*        may be abbreviated, to no less than 3 characters.
*     *value
*        Returned holding the value of the parameter.
*     *status
*        The global status.

*  Notes:
*     See the ndfTune function for a list of the tuning parameters
*     currently available.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     3-APR-2019 (DSB):
*        Original version, based on equivalent Fortran function by RFWS.
*     26-APR-2019 (DSB):
*        Add the FIXSW tuning parameter.
*     21-APR-2020 (DSB):
*        Add the ROUND tuning parameter.

*-
*/

/* Local Variables: */
   char xname[ DAT__SZNAM + 1 ];   /* An NDF extension name */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure the NDF library has been initialised. */
   NDF_INIT( status );
   if( *status == SAI__OK ) {

/* The TCB values are held in global variabl;es, so use a mutex to ensure
   the current thread has sole access. */
      NDF__TCB_LOCK_MUTEX;

/* Test the tuning parameter name supplied against each permitted value
   in turn, allowing abbreviation... */

/* Error tracing flag.
   ==================
   If TRACE was specified, then return the error tracing flag value. */
      if( ndf1Simlr( tpar, 1, 0, "TRACE", NDF__MINAB ) ) {
         if( Ndf_TCB_etflg ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Do format conversion flag.
   =========================
   If DOCVT was specified, then return the do format conversion flag
   value. */
      } else if( ndf1Simlr( tpar, 1, 0, "DOCVT", NDF__MINAB ) ) {
         if( Ndf_TCB_docvt ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Keep NDF objects flag.
   =====================
   If KEEP was specified, then return the keep NDF objects flag value. */
      } else if( ndf1Simlr( tpar, 1, 0, "KEEP", NDF__MINAB ) ) {
         if( Ndf_TCB_keep ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Create History component automatically flag.
   ============================================ */
      } else if( ndf1Simlr( tpar, 1, 0, "AUTO_HISTORY", NDF__MINAB ) ) {
         if( Ndf_TCB_autohistory ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Show data conversion flag.
   =========================
   If SHCVT was specified, then return the show data conversion flag
   value. */
      } else if( ndf1Simlr( tpar, 1, 0, "SHCVT", NDF__MINAB ) ) {
         if( Ndf_TCB_shcvt ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Warning message flag.
   ====================
   If WARN was specified, then return the warning message flag value. */
      } else if( ndf1Simlr( tpar, 1, 0, "WARN", NDF__MINAB ) ) {
         if( Ndf_TCB_warn ) {
            *value = 1;
         } else {
            *value = 0;
         }

/* Extension propagation.
   ======================
   Any tuning flag that begins with "PXT" is assumed to be terminated
   with the name of an extension. The tuning flag value is non-zero if
   the extension should be propagated by default by ndfProp, etc, and
   is zero if the extension should not be propagated by default. */
      } else if( ndf1Simlr( tpar, 0, 2, "PXT", NDF__MINAB ) ) {

/* copy the extension name, terminate it to remove trailing spaces and
   check it is not blank. */
         star_strlcpy( xname, tpar + 3, sizeof( xname ) );
         astChrTrunc( xname );
         if( *xname != 0 ) {

/* If there is a KeyMap, and it contains a value for the named extension,
   use the stored value. Otherwise use a defualt value of 1 (all
   extensions are propagated by default). */
            if( Ndf_TCB_pxt ) {
               if( !astMapGet0I( Ndf_TCB_pxt, xname, value ) ) *value = 1;
            } else {
               *value = 1;
            }

/* If the extension name was blank, then report an error. */
         } else {
            *status = NDF__TPNIN;
            errRep( " ", "'PXT' is not a valid tuning parameter name - it "
                    "should be followed by an NDF extension name (possible "
                    "programming error).", status );
         }

/* Maximum size of a section.
   ========================== */
      } else if( ndf1Simlr( tpar, 1, 0, "SECMAX", NDF__MINAB ) ) {
         *value = Ndf_TCB_secmax;

/* Fix the date/time strings stored in history records.
   =================================================== */
      } else if( ndf1Simlr( tpar, 1, 0, "FIXDT", NDF__MINAB ) ) {
         *value = Ndf_TCB_fixdt;

/* Fix the software strings stored in history records.
   ================================================== */
      } else if( ndf1Simlr( tpar, 1, 0, "FIXSW", NDF__MINAB ) ) {
         *value = Ndf_TCB_fixsw;

/* Round floating-point values to the nearest integer.
   ================================================== */
      } else if( ndf1Simlr( tpar, 1, 0, "ROUND", NDF__MINAB ) ) {
         *value = Ndf_TCB_round;

/* Unknown tuning parameter.
   ========================
   Report an error if the tuning parameter name is not recognised. */
      } else {
         *status = NDF__TPNIN;
         msgSetc( "TPAR", tpar );
         errRep( " ", "'^TPAR' is not a valid tuning parameter name "
                 "(possible programming error).", status );
      }

/* Release the mutex. */
      NDF__TCB_UNLOCK_MUTEX;
   }

/* If an error occurred, then report context information and call the
   error tracing function. */
   if( *status != SAI__OK ) {
      errRep( " ", "ndfGtune: Error obtaining the value of an NDF_ system "
              "tuning parameter.", status );
      ndf1Trace( "ndfGtune", status );
   }

/* Restablish the original AST status pointer */
   NDF_FINAL

}

